### PR TITLE
Bug 1773682: Add missing optional tags

### PR DESCRIPTION
**Pull request by soltysh**
/assign @deads2k @sttts 
One of the steps to land openshift/oc#149

I wonder how much we are still missing, our API's description is not consistent :disappointed:

### CODE REVIEW

File: template/v1/types.go [hunk 1]
```diff
@@ -111,6 +111,7 @@ type TemplateInstance struct {
 	Spec TemplateInstanceSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
 
 	// status describes the current state of this TemplateInstance.
+	// +optional
```
Comment (sttts) [inline template/v1/types.go]: do these have any effect, e.g. on the openshift-apiserver openapi spec?

File: apps/v1/types.go [hunk 2]
```diff
@@ -292,18 +292,18 @@ type DeploymentTriggerImageChangeParams struct {
 type DeploymentConfigStatus struct {
 	// LatestVersion is used to determine whether the current deployment associated with a deployment
 	// config is out of sync.
-	LatestVersion int64 `json:"latestVersion" protobuf:"varint,1,opt,name=latestVersion"`
+	LatestVersion int64 `json:"latestVersion,omitempty" protobuf:"varint,1,opt,name=latestVersion"`
```
Comment (smarterclayton) [inline apps/v1/types.go]: These should not be omit empty.  This would be a breaking change to JSON users, which have always set these fields.

**Comment by smarterclayton** [comment]
/hold

This changes the JSON serialization of very old APIs, and is not safe (particularly so for these APIS).

omit empty should not be required for optional status - if so that's a bug in our upstream code.

File: apps/v1/types.go (hunk 2 above)
Comment (sttts) [reply apps/v1/types.go]: @soltysh why not `// +optional` ?

**Comment by sttts** [comment]
/retitle Bug 1773682: Add missing optional tags and omitempty

**Comment by sttts** [comment]
@soltysh to solve @smarterclayton 's concerns about serialization, please split this into omitempty and `+optional` addition.

**Comment by scniro** [comment]
merge worthy?! I have been blocked moving forward using helm@3 to deploy routes. Hope to see this in very soon 🙏

**Comment by scniro** [comment]
FWIW I was able to move forward by stubbing out a junk `status` object

```yaml
status:
  ingress:
    - conditions:
      - lastTransitionTime: "2019-12-06T03:24:58Z"
        status: "True"
        type: Admitted
      host: <host>
      routerCanonicalHostname: <host>
      routerName: default
      wildcardPolicy: None
```

**Comment by sbose78** [comment]
The devtools team is blocked - need this in soon :)

File: apps/v1/types.go (hunk 2 above)
Comment (soltysh) [reply apps/v1/types.go]: I think we need `// +optional` but at the status level not here, looking at other types. Working on fixes.

File: template/v1/types.go (hunk 1 above)
Comment (soltysh) [reply template/v1/types.go]: Yes, they should not be required, and that's one of the reasons https://github.com/openshift/oc/pull/149 fails.

**Comment by soltysh** [comment]
@sttts @smarterclayton I've split the controversial `omitempty` changes into https://github.com/openshift/api/pull/538. This PR only adds `// +optional` where necessary. ptal 
With this PR having only optional tags I'm dropping hold, I've put it in the other PR, though, since that's where the controversy is. 
/hold cancel

File: route/v1/types.go [hunk 3]
```diff
@@ -122,6 +124,7 @@ type RouteTargetReference struct {
 	// weight as an integer between 0 and 256, default 100, that specifies the target's relative weight
 	// against other target reference objects. 0 suppresses requests to this backend.
 	Weight *int32 `json:"weight" protobuf:"varint,3,opt,name=weight"`
+	// +optional
```
Comment (sttts) [inline route/v1/types.go]: above?

File: template/v1/types.go [hunk 4]
```diff
@@ -126,6 +127,7 @@ type TemplateInstanceSpec struct {
 	// requester holds the identity of the agent requesting the template
 	// instantiation.
 	Requester *TemplateInstanceRequester `json:"requester" protobuf:"bytes,3,opt,name=requester"`
+	// +optional
```
Comment (sttts) [inline template/v1/types.go]: above?

**Comment by soltysh** [comment]
@sttts fixed

**Pull request merged**